README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# googler <img src='man/figures/logo.png' align="right" height="200" />

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/mkearney/googler.svg?branch=master)](https://travis-ci.org/mkearney/googler)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/mkearney/googler?branch=master&svg=true)](https://ci.appveyor.com/project/mkearney/googler)
<!-- badges: end -->

This is a wrapper around the [‘googler’ command line
tool](https://github.com/jarun/googler).

## Requirements

Python 3 must be installed and locatable

## Installation

You can install the development version of {googler} from
[Github](https://github.com/mkearney/googler) with:

``` r
remotes::install_github("mkearney/googler")
```

<!-- You can install the released version of googler from [CRAN](https://CRAN.R-project.org) with: -->

<!-- ``` r -->

<!-- install.packages("googler") -->

<!-- ``` -->

## Examples

Search google with various options to specify things like the desired
result **`count`**, google **`news`**, within a range of **`time`**,
from a given **`site`**, etc.

### Example \#1: Basic google search

``` r
## search google for 'rstats'
googler("rstats")
#> # A tibble: 10 x 4
#>    abstract                        matches    title            url                        
#>    <chr>                           <list>     <chr>            <chr>                      
#>  1 "5h ago @akesslerdc tweeted: \… <df[,2] [… News about #rst… https://twitter.com/hashta…
#>  2 The latest Tweets from Rstats … <df[,2] [… Rstats (@rstats… https://twitter.com/rstats…
#>  3 The R Project for Statistical … <NULL>     The R Project f… https://www.r-project.org/ 
#>  4 Read writing about Rstats in T… <df[,2] [… Rstats – Toward… https://towardsdatascience…
#>  5 Right now I'm building a GLM f… <NULL>     Rstats - Reddit  https://www.reddit.com/r/r…
#>  6 This function is used to calcu… <NULL>     rstats function… https://www.rdocumentation…
#>  7 Jun 24, 2019 ... GitHub is whe… <NULL>     rstats · GitHub… https://github.com/topics/…
#>  8 Read writing about Rstats in l… <df[,2] [… Rstats – learn … https://blog.exploratory.i…
#>  9 Jul 24, 2019 ... Welcome to RS… <df[,2] [… RStats Institut… https://www.missouristate.…
#> 10 Today I pushed a preliminary v… <NULL>     Rstats - Once U… https://www.onceupondata.c…
```

### Example \#2: Advanced google search

``` r
## search google for 100 'rstats' or "R language" hits in English in past year
googler("rstats OR \"R language\"", count = 100, time = "y1", lang = "en")
#> # A tibble: 97 x 5
#>    abstract                   matches   title               url                   metadata
#>    <chr>                      <list>    <chr>               <chr>                 <chr>   
#>  1 "5h ago @akesslerdc tweet… <df[,2] … News about #rstats… https://twitter.com/… <NA>    
#>  2 Right now I'm building a … <NULL>    Rstats - Reddit     https://www.reddit.c… <NA>    
#>  3 This past year in R has b… <NULL>    My 2018 #rstats ye… https://leonawicz.gi… <NA>    
#>  4 GitHub is where people bu… <NULL>    rstats · GitHub To… https://github.com/t… <NA>    
#>  5 Welcome to RStats Institu… <df[,2] … RStats Institute -… https://www.missouri… <NA>    
#>  6 RStats Glossary. Find def… <df[,2] … Stats Tools - RSta… https://www.missouri… <NA>    
#>  7 Dec 13, 2018 - 77 min - U… <df[,2] … R Programming for … https://www.youtube.… <NA>    
#>  8 Why the R Language? R is … <df[,2] … Why use the R Lang… https://www.burns-st… <NA>    
#>  9 "The view_df() function f… <NULL>    Quickly create Cod… https://www.r-blogge… <NA>    
#> 10 Basically, this book is a… <df[,2] … A Complete Guide t… https://towardsdatas… <NA>    
#> # … with 87 more rows
```

### Example \#3: Search google news

``` r
## search google news for rstats/R language articles in the past three weeks
googler("rstats OR \"R language\"", news = TRUE)
#> # A tibble: 10 x 5
#>    abstract                matches   metadata       title             url                 
#>    <chr>                   <list>    <chr>          <chr>             <chr>               
#>  1 Quartz: How would you … <df[,2] … Quartz, Aug 1… What's next for … https://qz.com/1661…
#>  2 ... R package develope… <df[,2] … Yahoo Finance… R Consortium Com… https://finance.yah…
#>  3 'Brittany Runs a Marat… <NULL>    The Boston Gl… Jillian Bell lac… https://www.bostong…
#>  4 “A broad range of orga… <df[,2] … Datanami, Aug… Is Python Strang… https://www.datanam…
#>  5 ... data visualization… <df[,2] … Neowin, Aug 3… This 2019 Certif… https://www.neowin.…
#>  6 28 R (language) — Gary… <df[,2] … Valdosta Dail… Capsule reviews … https://www.valdost…
#>  7 The R language offers … <df[,2] … Toolbox (blog… Google's AI Plat… https://it.toolbox.…
#>  8 Her favorites are Data… <df[,2] … Quartz, Nov 1… What's the best … https://qz.com/1464…
#>  9 Big-data company Datab… <NULL>    SiliconANGLE … Databricks' MLfl… https://siliconangl…
#> 10 The R language was ori… <df[,2] … IT Jungle, Ap… R Comes To i      https://www.itjungl…
```


NEWS.md
# googler 0.0.1

* Added a `NEWS.md` file to track changes to the package.
* Initial build


tests/testthat.R
library(testthat)
library(googler)

test_check("googler")


tests/testthat/test-googler.R
test_that("googler works", {
  ## if python 3 not found then expect error
  #if (!googler:::py3_installed()) {
    #expect_error(googler("hello", count = 1))
  #} else {
    x <- googler("hello", count = 20)
    expect_true(is.data.frame(x))
    #expect_true(nrow(x) > 1)
    #expect_true(all(c("url", "abstract") %in% names(x)))
  #}
})


R/path.R

try_null <- function(x) {
  tryCatch(eval(x, envir = parent.frame()), error = function(e) NULL)
}
try_tbl <- function(x) {
  tryCatch(eval(x, envir = parent.frame()), error = function(e) tibble::tibble())
}

ssystem <- function(...,
                    intern = TRUE,
                    ignore.stdout = FALSE,
                    ignore.stderr = FALSE,
                    quiet = FALSE) {
  if (quiet) {
    ignore.stderr <- TRUE
  }
  try_null(system(paste0(c(...), collapse = ""),
    intern = intern,
    ignore.stdout = ignore.stdout,
    ignore.stderr = ignore.stderr))
}


py3_installed <- function() {
  isTRUE(grepl("python", py_path(), ignore.case = TRUE))
}

py_path <- function() {
  ## path to conda
  conda <- list.dirs("~", full.names = TRUE, recursive = FALSE)
  conda <- paste0(grep("conda3", conda, value = TRUE), "/bin")

  ## system paths
  sp <- Sys.getenv("PATH")
  if (sp == "") {
    sp <- "/usr/bin"
  }
  sys_paths <- strsplit(sp, "/?:")[[1]]

  ## combine with conda
  sys_paths <- unique(c(sys_paths, conda))

  ## look for python executables
  py_paths <- unique(c(unlist(lapply(sys_paths, function(.x)
    list.files(.x, pattern = "python($|.*[^mg]$)",
    ignore.case = TRUE, full.names = TRUE))),
    sys_which(paste0("python",
      c("", "4", "3.9", "3.8", "3.7", "3.6", "3.5", "3.4", "3")))))

  ## with the number 3 in them
  py_paths <- grep("3", py_paths, value = TRUE)

  ## if no python return empty
  if (length(py_paths) == 0) {
    return("")
  }

  ## if 3.4 or above
  if (any(grepl("python.*3\\.[4-9]", py_paths, ignore.case = TRUE))) {
    py_paths <- grep("python.*3\\.[4-9]", py_paths, ignore.case = TRUE, value = TRUE)
    return(py_paths[1])
  }

  ## if python3
  if (any(grepl("python.*3", py_paths, ignore.case = TRUE))) {
    py_paths <- grep("python.*3", py_paths, ignore.case = TRUE, value = TRUE)
    return(py_paths[1])
  }

  ## otherwise return normal python
  return(py_paths[1])
}


R/googler.R
#' googler
#'
#' googler(): Search Google from the R console
#'
#' @param query Search query.
#' @param start Start at the Nth result.
#' @param count Show N results–defaults to 10
#' @param news Show results from news section.
#' @param tld Country-specific search with top-level domain TLD, e.g., 'in' for India.
#' @param lang Search for the language, e.g., 'fi' for Finnish.
#' @param exact Disable automatic spelling correction. Search exact keywords.
#' @param time Time limit search, e.g., 'h5' (5 hrs), 'd5' (5 days), 'w5' (5 weeks), 'm5' (5 months), 'y5' (5 years)
#' @param site Search a site using Google.
#' @param unfilter Do not omit similar results.
#' @return A tibble data frame where each row represents one search result and columns contain information about the abstract, 
#'   metadata, url, and phrase matching for each result.
#' @examples
#' \dontrun{
#' ## search google within past month
#' googler("rstats", time = "m1")
#'
#' ## search google news
#' googler_news("rstats")
#' }
#' @export
googler <- function(query,
                    start = NULL,
                    count = NULL,
                    news = NULL,
                    tld = NULL,
                    lang = NULL,
                    exact = NULL,
                    time = NULL,
                    site = NULL,
                    unfilter = NULL) {
  ## parse JSON and return as a tibble
  try_tbl(tibble::as_tibble(jsonlite::fromJSON(
    do.call("googler_", as.list(environment()))
  )))
}

#' googler_
#'
#' googler_(): Internal version of googler that returns stdout
#'
#' @inheritParams googler
#' @keywords internal
googler_ <- function(query,
                     start = NULL,
                     count = NULL,
                     news = NULL,
                     tld = NULL,
                     lang = NULL,
                     exact = NULL,
                     time = NULL,
                     site = NULL,
                     unfilter = NULL) {

  ## path to python3
  cmd <- py_path()

  ## path to googler
  ggl <- googler_path()

  ## issue automation warning if count > 100
  check_count(count > 100)

  ## check query
  check_query(query)

  ## set/reset encoding
  op <- getOption("encoding")
  on.exit(options(encoding = op), add = TRUE)
  options(encoding = "UTF-8")

  ## compile args
  args <- c(ggl, shQuote(query),
    start    %|||% paste0("--start ", start),
    count    %|||% paste0("--count ", count),
    news     %|||% paste0("--news"),
    tld      %|||% paste0("--tld ", tld),
    lang     %|||% paste0("--lang ", lang),
    exact    %|||% paste0("--exact"),
    time     %|||% paste0("--time ", time),
    site     %|||% paste0("--site ", site),
    unfilter %|||% paste0("--unfilter"),
    "-C", "--json"
  )

  ## execute command and return stdout
  system2(cmd, args = args, stdout = TRUE)
}

python_path <- function() {
  path <- sys_which(paste0("python",
    c("", "4", "3.9", "3.8", "3.7", "3.6", "3.5", "3.4", "3")))
  if (!any(grepl("python[[:punct:] ]{0,}3|3[[:punct:] ]{0,}python",
    path, ignore.case = TRUE))) {
    path <- grep("python", path, value = TRUE, ignore.case = TRUE)[1]
  } else {
    path <- grep("3[[:punct:] ]{0,}python|python[[:punct:] ]{0,}3",
      path, value = TRUE, ignore.case = TRUE)[1]
  }
  if (is.na(path) || !nzchar(path)) {
    stop("'googler' requires python3, which does not appear to be installed.")
  }
  path
}

is_unix <- function() grepl("unix", .Platform$OS.type, ignore.case = TRUE)

win_py_version <- function() {
  system("python --version", intern = TRUE, invisible = TRUE)
}

sys_which <- function(names) {
  names <- names[!is.na(names)]
  res <- character(length(names))
  names(res) <- names
  if (is_unix()) {
    which <- "/usr/bin/which"
  } else {
    which <- "where"
  }
  for (i in seq_along(names)) {
    ans <- suppressWarnings(
      tryCatch(system(paste(which, shQuote(names[i])),
      intern = TRUE, ignore.stderr = TRUE),
        error = function(e) ""))
    if (grepl("solaris", R.version$os)) {
      tmp <- strsplit(ans[1], " ", fixed = TRUE)[[1]]
      if (identical(tmp[1:3], c("no", i, "in")))
        ans <- ""
    }
    res[i] <- if (length(ans))
      ans[1]
    else ""
    if (is.na(res[i]))
      res[i] <- ""
    if (!file.exists(res[i]))
      res[i] <- ""
  }
  res
}

check_count <- function(x) {
  if (length(x) == 0) {
    return(invisible())
  }
  if (x) {
    warning("Automated bulk web scraping is not the intended use of {googler}",
      immediate. = TRUE)
  }
  invisible()
}

check_query <- function(query) {
  stopifnot(
    length(query) == 1,
    is.character(query),
    nzchar(query)
  )
}

`%|||%` <- function(a, b) {
  if (!is.null(a))
    b
  else
    a
}

googler_path <- function() {
  ## if googler.path in options
  if (!is.null(path <- getOption("googler.path"))) {
    return(path)
  }
  ## if googler is installed
  if (grepl("googler", path <- sys_which("googler"))) {
    options(googler.path = path)
    return(path)
  }

  ## otherwise download and store script
  if (file.exists(file.path(tempdir(), "googler"))) {
    unlink(file.path(tempdir(), "googler"))
  }
  utils::download.file(
    "https://raw.githubusercontent.com/jarun/googler/v3.9/googler",
    path <- file.path(tempdir(), "googler"),
    quiet = TRUE
  )
  options(googler.path = path)
  path
}
